File: OnlinePK-Android/biz-live/src/main/java/com/netease/biz_live/yunxin/live/dialog/adapter/AudienceListAdapter.java
/*
 * Copyright (c) 2021 NetEase, Inc.  All rights reserved.
 * Use of this source code is governed by a MIT license that can be found in the LICENSE file.
 */

package com.netease.biz_live.yunxin.live.dialog.adapter;

import android.app.Activity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.blankj.utilcode.util.ToastUtils;
import com.netease.biz_live.R;
import com.netease.biz_live.yunxin.live.dialog.ChoiceDialog;
import com.netease.biz_live.yunxin.live.dialog.fragment.AudienceListFragment;
import com.netease.biz_live.yunxin.live.liveroom.AnchorSeatManager;
import com.netease.biz_live.yunxin.live.liveroom.LiveRoomCallback;
import com.netease.biz_live.yunxin.live.liveroom.NERTCLiveRoom;
import com.netease.biz_live.yunxin.live.model.SeatMemberInfo;
import com.netease.biz_live.yunxin.live.utils.ClickUtils;
import com.netease.yunxin.android.lib.picture.ImageLoader;

import java.util.ArrayList;
import java.util.List;

/**
 * 观众列表
 */
public class AudienceListAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private ArrayList<SeatMemberInfo> memberInfos;

    private int type;

    private Activity context;

    private AnchorSeatManager seatManager;

    public AudienceListAdapter(Activity context, int type) {
        memberInfos = new ArrayList<>();
        this.type = type;
        this.context = context;
        seatManager = NERTCLiveRoom.sharedInstance(true).getService(AnchorSeatManager.class);
    }

    public void setData(List<SeatMemberInfo> members){
        if(members == null){
            return;
        }
        memberInfos.clear();
        memberInfos.addAll(members);
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        switch (viewType){
            case AudienceListFragment.TYPE_APPLY:
                View apply = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_item_audience_apply, parent, false);
                return new ApplyAudienceViewHolder(apply);
            case AudienceListFragment.TYPE_MANAGER:
                View seat = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_item_audience_seats, parent, false);
                return new SeatAudienceViewHolder(seat);
            default:
                View common = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_item_audience_common, parent, false);
                return new AudienceCommonViewHolder(common);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (context == null) {
            return;
        }
        if (holder instanceof AudienceViewHolder) {
            SeatMemberInfo member = memberInfos.get(position);
            ((AudienceViewHolder) holder).mTvNumber.setText(String.valueOf(position + 1));
            ((AudienceViewHolder) holder).mTvNick.setText(member.nickName);
            ImageLoader.with(context).circleLoad(member.avatar, ((AudienceViewHolder) holder).mIvAvatar);
            if (holder instanceof AudienceCommonViewHolder) {
                ((AudienceCommonViewHolder) holder).mTvInvite.setOnClickListener(v -> {
                    if (ClickUtils.isFastClick()) {
                        return;
                    }
                    if (seatManager != null) {
                        seatManager.pickSeat(member.accountId, new LiveRoomCallback<Void>() {
                            @Override
                            protected void onSuccess() {
                                super.onSuccess();
                                removeMember(member);
                                ToastUtils.showShort(R.string.anchor_invite_success);
                            }

                            @Override
                            public void onError(int code, String msg) {
                                ToastUtils.showShort(msg);
                            }
                        });
                    }
                });
            }else if(holder instanceof ApplyAudienceViewHolder){
                ((ApplyAudienceViewHolder) holder).mTvAccept.setOnClickListener(v -> {
                    if (ClickUtils.isFastClick()) {
                        return;
                    }
                    if (seatManager != null) {
                        seatManager.acceptSeatApply(member.accountId, new LiveRoomCallback<Void>() {
                            @Override
                            protected void onSuccess() {
                                super.onSuccess();
                                removeMember(member);
                                ToastUtils.showShort(R.string.have_accept);
                            }

                            @Override
                            public void onError(int code, String msg) {
                                ToastUtils.showShort(msg);
                            }
                        });
                    }
                });
                ((ApplyAudienceViewHolder) holder).mTvReject.setOnClickListener(v -> {
                    if (ClickUtils.isFastClick()) {
                        return;
                    }
                    if (seatManager != null) {
                        seatManager.rejectSeatApply(member.accountId, new LiveRoomCallback<Void>() {
                            @Override
                            protected void onSuccess() {
                                super.onSuccess();
                                removeMember(member);
                                ToastUtils.showShort(R.string.have_reject);
                            }

                            @Override
                            public void onError(int code, String msg) {
                                ToastUtils.showShort(msg);
                            }
                        });
                    }
                });
            }else if(holder instanceof  SeatAudienceViewHolder){
                ((SeatAudienceViewHolder) holder).mIvAudio.setSelected(member.audio == 0);
                ((SeatAudienceViewHolder) holder).mIvVideo.setSelected(member.video == 0);
                ((SeatAudienceViewHolder) holder).mIvAudio.setOnClickListener(v -> {
                    if (ClickUtils.isFastClick()) {
                        return;
                    }
                    if (seatManager != null) {
                        seatManager.setSeatMuteState(member.accountId, v.isSelected() ? 1 : 0, member.video, new LiveRoomCallback<Void>() {
                            @Override
                            protected void onSuccess() {
                                super.onSuccess();
                                v.setSelected(!v.isSelected());
                                member.audio = v.isSelected() ? 0 : 1;
                            }

                            @Override
                            public void onError(int code, String msg) {
                                ToastUtils.showShort(msg);
                            }
                        });
                    }
                });
                ((SeatAudienceViewHolder) holder).mIvVideo.setOnClickListener(v -> {
                    if (ClickUtils.isFastClick()) {
                        return;
                    }
                    if (seatManager != null) {
                        seatManager.setSeatMuteState(member.accountId, member.audio, v.isSelected() ? 1 : 0, new LiveRoomCallback<Void>() {
                            @Override
                            protected void onSuccess() {
                                super.onSuccess();
                                v.setSelected(!v.isSelected());
                                member.video = v.isSelected() ? 0 : 1;
                            }

                            @Override
                            public void onError(int code, String msg) {
                                ToastUtils.showShort(msg);
                            }
                        });
                    }
                });
                ((SeatAudienceViewHolder) holder).mTvHangup.setOnClickListener(v -> {
                    if (ClickUtils.isFastClick()) {
                        return;
                    }
                    showKickDialog(member);
                });
            }
        }
    }

    private void removeMember(SeatMemberInfo member) {
        int index = memberInfos.indexOf(member);
        if (index >= 0) {
            memberInfos.remove(index);
            notifyItemRemoved(index);
        }
    }

    /**
     * 踢下麦二次确认
     *
     * @param member
     */
    private void showKickDialog(SeatMemberInfo member) {
        if (context == null) {
            return;
        }
        ChoiceDialog dialog = new ChoiceDialog(context);
        dialog.setContent(String.format("是否挂断与%s的连麦？", member.nickName))
                .setNegative("取消", v -> {
                })
                .setPositive("挂断", v -> {
                    if (seatManager != null) {
                        seatManager.kickSeat(member.accountId, new LiveRoomCallback<Void>() {

                            @Override
                            protected void onSuccess() {
                                super.onSuccess();
                                removeMember(member);
                                ToastUtils.showShort(R.string.have_leave_seat);
                            }

                            @Override
                            public void onError(int code, String msg) {
                                ToastUtils.showShort(msg);
                            }
                        });
                    }
                }).show();
    }

    @Override
    public int getItemViewType(int position) {
        return type;
    }

    @Override
    public int getItemCount() {
        return memberInfos.size();
    }

    private static class AudienceViewHolder extends RecyclerView.ViewHolder{

        public TextView mTvNumber;

        public ImageView mIvAvatar;

        public TextView mTvNick;

        public AudienceViewHolder(@NonNull View itemView) {
            super(itemView);
            mTvNumber = itemView.findViewById(R.id.tv_audience_no);
            mIvAvatar = itemView.findViewById(R.id.iv_audience_avatar);
            mTvNick = itemView.findViewById(R.id.tv_audience_nickname);
        }
    }

    private static class AudienceCommonViewHolder extends AudienceViewHolder{

        public TextView mTvInvite;

        public AudienceCommonViewHolder(@NonNull View itemView) {
            super(itemView);
            mTvInvite = itemView.findViewById(R.id.tv_invite);
        }
    }

    private static class ApplyAudienceViewHolder extends AudienceViewHolder{

        public TextView mTvReject;

        public TextView mTvAccept;

        public ApplyAudienceViewHolder(@NonNull View itemView) {
            super(itemView);
            mTvAccept = itemView.findViewById(R.id.tv_accept);

            mTvReject = itemView.findViewById(R.id.tv_reject);
        }
    }

    private static class SeatAudienceViewHolder extends AudienceViewHolder{

        public ImageView mIvVideo;

        public ImageView mIvAudio;

        public TextView mTvHangup;

        public SeatAudienceViewHolder(@NonNull View itemView) {
            super(itemView);
            mIvVideo = itemView.findViewById(R.id.iv_video);

            mIvAudio = itemView.findViewById(R.id.iv_audio);

            mTvHangup = itemView.findViewById(R.id.tv_hangup);
        }
    }
}

File: OnlinePK-Android/biz-live/src/main/java/com/netease/biz_live/yunxin/live/anchor/dialog/adapter/AudienceListAdapter.kt
/*
 *  Copyright (c) 2021 NetEase, Inc.  All rights reserved.
 *  Use of this source code is governed by a MIT license that can be found in the LICENSE file
 */

package com.netease.biz_live.yunxin.live.anchor.dialog.adapter

import android.app.Activity
import android.view.*
import android.widget.*
import androidx.recyclerview.widget.RecyclerView
import com.blankj.utilcode.util.*
import com.netease.biz_live.R
import com.netease.biz_live.yunxin.live.dialog.ChoiceDialog
import com.netease.biz_live.yunxin.live.utils.ClickUtils
import com.netease.yunxin.android.lib.picture.ImageLoader
import com.netease.yunxin.seatlibrary.CompletionCallback
import com.netease.yunxin.seatlibrary.seat.constant.AudienceType
import com.netease.yunxin.seatlibrary.seat.constant.SeatAVState
import com.netease.yunxin.seatlibrary.seat.model.SeatInfo
import com.netease.yunxin.seatlibrary.seat.params.AnchorActionSeatParams
import com.netease.yunxin.seatlibrary.seat.params.SetSeatAVMuteStateParams
import com.netease.yunxin.seatlibrary.seat.service.SeatService
import java.util.*

/**
 * 观众列表
 */
class AudienceListAdapter(context: Activity?, type: Int) :
    RecyclerView.Adapter<RecyclerView.ViewHolder?>() {
    private val memberInfos: ArrayList<SeatInfo> = ArrayList()
    private val type: Int = type
    private val context: Activity? = context
    private val seatService by lazy { SeatService.sharedInstance() }

    fun setData(members: MutableList<SeatInfo>?) {
        if (members == null) {
            return
        }
        memberInfos.clear()
        memberInfos.addAll(members)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            AudienceType.AUDIENCE_TYPE_APPLY -> {
                val apply = LayoutInflater.from(parent.context)
                    .inflate(R.layout.view_item_audience_apply, parent, false)
                ApplyAudienceViewHolder(apply)
            }
            AudienceType.AUDIENCE_TYPE_ON_SEAT -> {
                val seat = LayoutInflater.from(parent.context)
                    .inflate(R.layout.view_item_audience_seats, parent, false)
                SeatAudienceViewHolder(seat)
            }
            else -> {
                val common = LayoutInflater.from(parent.context)
                    .inflate(R.layout.view_item_audience_common, parent, false)
                AudienceCommonViewHolder(common)
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (context == null) {
            return
        }
        if (holder is AudienceViewHolder) {
            val member = memberInfos.get(position)
            holder.mTvNumber.text = (position + 1).toString()
            holder.mTvNick.text = member.nickName
            ImageLoader.with(context)
                .circleLoad(member.avatar, holder.mIvAvatar)
            val params = AnchorActionSeatParams(null, member.accountId)
            when (holder) {
                is AudienceCommonViewHolder -> {
                    holder.mTvInvite.setOnClickListener(View.OnClickListener {
                        if (ClickUtils.isFastClick()) {
                            return@OnClickListener
                        }
                        seatService.pickSeat(params, object : CompletionCallback<String> {
                            override fun success(info: String?) {
                                removeMember(member)
                                ToastUtils.showShort(R.string.biz_live_anchor_invite_success)
                            }

                            override fun error(code: Int, msg: String) {
                                ToastUtils.showShort(msg)
                            }


                        })
                    })
                }
                is ApplyAudienceViewHolder -> {
                    holder.mTvAccept.setOnClickListener(View.OnClickListener {
                        if (ClickUtils.isFastClick()) {
                            return@OnClickListener
                        }
                        seatService.acceptSeatApply(
                            params,
                            object : CompletionCallback<Void> {
                                override fun success(info: Void?) {
                                    removeMember(member)
                                    ToastUtils.showShort(R.string.biz_live_have_accepted)
                                }

                                override fun error(code: Int, msg: String) {
                                    ToastUtils.showShort(msg)
                                }
                            })
                    })
                    holder.mTvReject.setOnClickListener(View.OnClickListener { v: View? ->
                        if (ClickUtils.isFastClick()) {
                            return@OnClickListener
                        }
                        seatService.rejectSeatApply(
                            params,
                            object : CompletionCallback<Void> {
                                override fun success(info: Void?) {
                                    removeMember(member)
                                    ToastUtils.showShort(R.string.biz_live_have_reject)
                                }

                                override fun error(code: Int, msg: String) {
                                    ToastUtils.showShort(msg)
                                }
                            })
                    })
                }
                is SeatAudienceViewHolder -> {
                    holder.mIvAudio.isSelected = member.audioState == 0
                    holder.mIvVideo.isSelected = member.videoState == 0
                    holder.mIvAudio.setOnClickListener(View.OnClickListener { v: View ->
                        if (ClickUtils.isFastClick()) {
                            return@OnClickListener
                        }
                        val audioParams = SetSeatAVMuteStateParams(
                            null,
                            member.accountId,
                            if (v.isSelected) SeatAVState.OPEN else SeatAVState.CLOSE
                        )
                        seatService.setSeatAudioMuteState(
                            audioParams,
                            object : CompletionCallback<Void> {
                                override fun success(info: Void?) {
                                    v.isSelected = !v.isSelected
                                    member.audioState = if (v.isSelected) 0 else 1
                                }

                                override fun error(code: Int, msg: String) {
                                    ToastUtils.showShort(msg)
                                }
                            })
                    })
                    holder.mIvVideo.setOnClickListener(View.OnClickListener { v: View ->
                        if (ClickUtils.isFastClick()) {
                            return@OnClickListener
                        }
                        val videoParams = SetSeatAVMuteStateParams(
                            null,
                            member.accountId,
                            if (v.isSelected) SeatAVState.OPEN else SeatAVState.CLOSE
                        )
                        seatService.setSeatVideoMuteState(
                            videoParams,
                            object : CompletionCallback<Void> {
                                override fun success(info: Void?) {
                                    v.isSelected = !v.isSelected
                                    member.videoState = if (v.isSelected) 0 else 1
                                }

                                override fun error(code: Int, msg: String) {
                                    ToastUtils.showShort(msg)
                                }
                            })
                    })
                    holder.mTvHangup.setOnClickListener(View.OnClickListener { v: View ->
                        if (ClickUtils.isFastClick()) {
                            return@OnClickListener
                        }
                        showKickDialog(member)
                    })
                }
            }
        }
    }

    private fun removeMember(member: SeatInfo?) {
        val index = memberInfos.indexOf(member)
        if (index >= 0) {
            memberInfos.removeAt(index)
            notifyItemRemoved(index)
        }
    }

    /**
     * 踢下麦二次确认
     *
     * @param member
     */
    private fun showKickDialog(member: SeatInfo?) {
        if (context == null) {
            return
        }
        val dialog = ChoiceDialog(context)
        dialog.setContent(
            String.format(
                Utils.getApp().getString(R.string.biz_live_sure_hanup_link_seat),
                member?.nickName
            )
        )
            .setNegative(Utils.getApp().getString(R.string.biz_live_cancel)) { }
            .setPositive(Utils.getApp().getString(R.string.biz_live_hangup)) {
                val params = AnchorActionSeatParams(null, member?.accountId)
                seatService.kickSeat(params, object : CompletionCallback<Void> {
                    override fun success(info: Void?) {
                        removeMember(member)
                        ToastUtils.showShort(R.string.biz_live_have_kick_seat)
                    }

                    override fun error(code: Int, msg: String) {
                        ToastUtils.showShort(msg)
                    }
                })
            }.show()
    }

    override fun getItemViewType(position: Int): Int {
        return type
    }

    override fun getItemCount(): Int {
        return memberInfos.size
    }

    private open class AudienceViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var mTvNumber: TextView = itemView.findViewById(R.id.tv_audience_no)
        var mIvAvatar: ImageView = itemView.findViewById(R.id.iv_audience_avatar)
        var mTvNick: TextView = itemView.findViewById(R.id.tv_audience_nickname)

    }

    private class AudienceCommonViewHolder(itemView: View) : AudienceViewHolder(itemView) {
        var mTvInvite: TextView = itemView.findViewById(R.id.tv_invite)

    }

    private class ApplyAudienceViewHolder(itemView: View) : AudienceViewHolder(itemView) {
        var mTvReject: TextView = itemView.findViewById(R.id.tv_reject)
        var mTvAccept: TextView = itemView.findViewById(R.id.tv_accept)

    }

    private class SeatAudienceViewHolder(itemView: View) : AudienceViewHolder(itemView) {
        var mIvVideo: ImageView = itemView.findViewById(R.id.iv_video)
        var mIvAudio: ImageView = itemView.findViewById(R.id.iv_audio)
        var mTvHangup: TextView = itemView.findViewById(R.id.tv_hangup)

    }

}